File: longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = {}
        for cur in nums:
            cnt = 0
            for prev in dp:
                if prev < cur and dp[prev] > cnt:
                    cnt = dp[prev]
            dp[cur] = 1 + cnt
        return max(dp.values())

File: after-camp/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> lis;
        for(auto n: nums){
            if(lis.empty() || lis.back() < n) lis.push_back(n);
            else{
                *lower_bound(lis.begin(), lis.end(), n) = n;
            }
        }
        return lis.size();
    }
};